### PR TITLE
feat: migrate to vty-crossplatform and brick-2

### PR DESCRIPTION
**Pull request by ShrykeWindgrace**
Addresses #2112

Incidentally, this adds support for `hledger-ui` for windows in Windows Terminal.

Current state of support of various features in different terminal emulators on Windows is tracked here: https://github.com/chhackett/vty-windows/issues/15 with visualization here: https://github.com/chhackett/vty-windows/wiki/TerminalSupport (WIP). Rule of thumb - if emulator is mintty-based (defaults for cygwin, git bash, msys2), the app will fail. If it is conhost-based (Windows-Terminal, Wezterm) the app will work. If its a conhost-based-but-opinionated (Alacritty), almost everything will work.

I have not yet started proper documentation for `hledger-ui`, but I think that in most cases it will be safe to redirect them to `vty-windows` outright.

Also I have not touched CI/releases/packaging. The builds should pass for sufficiently recent GHC (I think ghc-9.4.4 should be enough. ghc-8.* most probably won't work).

### CODE REVIEW

**Comment by ShrykeWindgrace** [comment]
@simonmichael I did not take that aspect into account, to be honest. I know that `scoop` and `chocolatey` for windows just grab the latest released executables from github, so they do not care. Nix and its derivatives most probably compile on whatever default `ghc` they have in nixpkgs (was `9.4.8` yesterday morning). For more mainstream *nix distros (Ubuntu, Debian, Arch, ...) I have zero idea 😥

**Review by simonmichael — COMMENTED** [review]
What docs changes do you think are needed ? I think not much (I can help with that if needed).

I'm willing to require GHC 9.4 if the poll I just sent out doesn't turn up problems.

Looking good !

**Comment by ShrykeWindgrace** [comment]
> What docs changes do you think are needed ? I think not much (I can help with that if needed).

Just mention that `hledger-ui` would not work in all terminal emulators people use on Windows.
 
> I'm willing to require GHC 9.4 if the poll I just sent out doesn't turn up problems.

I edited my previous comment. It seems that would not be necessary for packaging on non-windows OSes.

> Looking good !

Thanks! I'll iron out the extra-deps and dependencies for non-windows OSes soon(ish)=)

**Comment by simonmichael** [comment]
It looks like the right fix might be to add the newer packages in stack9.4.yaml.?

**Comment by Vekhir** [comment]
Hi, I'm in the process of updating the Arch packages to GHC 9.6 and `hledger-ui` is one of them. May I ask what is holding up this PR?

For my part, I can confirm that this patch works for Arch. I also tested with `brick` 2.2, which was released yesterday, and that works too. We are not using `stack` though.

-- Vekhir

**Comment by ShrykeWindgrace** [comment]
> Hi, I'm in the process of updating the Arch packages to GHC 9.6 and `hledger-ui` is one of them. May I ask what is holding up this PR?

Hi! Thanks for maintaining that package=) From my part - nothing software-related, just plain old lack of time because of end-of-the-year festivities and day job workload=) I'll try to move the PR forward this week.

**Comment by ShrykeWindgrace** [comment]
> It looks like the right fix might be to add the newer packages in stack9.4.yaml.?

Just did that=)

File: hledger-ui/package.yaml [hunk 1]
```diff
@@ -101,10 +98,13 @@ library:
   - time >=1.5
   - transformers
   - vector
-  # not installable on windows, cf buildable flag below
-  - brick >=1.5 && <2
-  - vty >=5.15 && <6
-  - unix
+  - brick >=2.1.1 && <3
+  - vty >=6.1 && <7
```
[the remark this answers is not in the document]
Comment (simonmichael) [reply hledger-ui/package.yaml]: Still need these tighter upper bounds, though I can do it as a fixup.

**Review by simonmichael — COMMENTED** [review]
@ShrykeWindgrace looking good, let's add the tighter upper bounds requested, and let's clarify the situation with minimum GHC requirement: I think we found requiring GHC 9.4+ isn't strictly necessary, but as it stands this PR makes hledger require GHC 9.4+, and I'm ok with trying this, but I wonder what's the experience now when you try to build with an older GHC. Shouldn't we bump the lower base bound in the hledger packages also to make our GHC requirement explicit.

File: hledger-ui/package.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vekhir) [reply hledger-ui/package.yaml]: At least on Arch, `brick 2.2` also works, so perhaps a bound of `< 2.3` is also appropriate for `brick`. On Stackage, it's available in Stackage Nightly 2023-12-19.

**Comment by simonmichael** [comment]
https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history has the GHC/base versions. base 4.17.0.0 allows GHC 9.4.0-4, we could require base 4.17.1.0 (GHC 9.4.5+) if it ensures a clearer build failure for installers.

**Comment by ShrykeWindgrace** [comment]
> @ShrykeWindgrace looking good, 

Thanks!

> let's add the tighter upper bounds requested,

I thought I did that. What did I miss?

> and let's clarify the situation with minimum GHC requirement: I think we found requiring GHC 9.4+ isn't strictly necessary, but as it stands this PR makes hledger require GHC 9.4+, and I'm ok with trying this, but I wonder what's the experience now when you try to build with an older GHC. 

If I understand correctly, this PR makes a ghc-9.4+ a requirement only on windows. Other OSes should not be impacted.

> Shouldn't we bump the lower base bound in the hledger packages also to make our GHC requirement explicit.

I do not have a good vision on how people build and package `hledger` on different OSes and platforms. On windows they usually grab an executable from the latest GH release (and hence don't care about GHC), on `nix` they build with their current GHC (9.4.8 at the moment); you mentioned that Ubuntus/Debians are lagging behind in their packaging (current Debian `testing` is 9.4.7, https://packages.debian.org/search?keywords=ghc), but then again, I don't know whether people use their pacman/yum/apt/brew/whatever or build from source to get a `hledger` executable. Personally, I will not feel this GHC version change at all; other people - I do not have enough info. Mailing list did not bring much objection, if I memory serves me right. Sorry of being no help here=)

File: hledger-ui/package.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShrykeWindgrace) [reply hledger-ui/package.yaml]: brick-2.2 is an enhancement over 2.1; judging by the changelog, there are no API changes.

[the remark this answers is not in the document]
Comment (ShrykeWindgrace) [reply hledger-ui/package.yaml]: > API changes are not unknown in these packages, I think we'd better put tighter upper bounds (~<2.2~ <2.3, <6.2) even though it's a hassle.

Already did that for 2.2/6.2, can bump brick to <2.3 if you want.

**Comment by simonmichael** [comment]
Ok, I'm always trying to smooth the haskell build experience as much as we can, but I'm just going to merge this and we'll see how it shakes out. Thank you very much !

**Pull request merged**